File: _dsv/flip.py
import argparse
from ._base import _Base
from . import _utils

class flip(_Base):
    ''' prints each column on a separate line '''
    parser = argparse.ArgumentParser()
    parser.add_argument('-n', '--lines', type=int, metavar='NUM', help='print the first NUM lines')
    parser.add_argument('--row-sep', choices=('never', 'always', 'auto'), default='auto', help='show a separator between the rows')
    parser.set_defaults(ofs=_Base.PRETTY_OUTPUT)

    def __init__(self, opts):
        super().__init__(opts)
        self.opts.row_sep = _utils.resolve_tty_auto(self.opts.row_sep)
        self.count = 0

    def on_header(self, header):
        header = [b'row', b'column']
        if self.header is not None:
            header.append(b'key')
        header.append(b'value')
        return super().on_header(header)

    def on_row(self, row):
        if self.count == 0:
            # first row
            if self.header is None:
                if self.on_header(None):
                    return True

        elif self.opts.row_sep:
            if super().on_row([b'---']):
                return True

        self.count += 1

        for i, value in enumerate(row, 1):
            row = [b'%i' % self.count, b'%i' % i]
            if self.header is not None:
                row.append(self.header[i-1] if i <= len(self.header) else b'')
            row.append(value)
            if super().on_row(row):
                return True

        if self.opts.lines and self.count >= self.opts.lines:
            return True

File: src/subcommands/flip.rs
use crate::utils::Break;
use anyhow::Result;
use crate::base;
use bstr::{BString};
use clap::{Parser};

#[derive(Parser)]
#[command(about = "prints each column on a separate line")]
pub struct Opts {
    #[arg(short = 'n', long, value_name = "NUM", help = "print the first NUM lines")]
    lines: Option<usize>,
    #[arg(long, value_enum, default_value_t = base::AutoChoices::Auto, help = "show a separator between the rows")]
    row_sep: base::AutoChoices,
}

pub struct Handler {
    opts: Opts,
    count: usize,
    header: Option<Vec<BString>>,
}

impl Handler {
    pub fn new(mut opts: Opts, base: &mut base::Base) -> Result<Self> {
        opts.row_sep = opts.row_sep.resolve(base.opts.inner.is_stdout_tty);
        if base.opts.ofs.is_none() {
            base.opts.pretty = true;
        }

        Ok(Self{
            opts,
            count: 0,
            header: None,
        })
    }
}

impl base::Processor for Handler {
    fn on_header(&mut self, base: &mut base::Base, header: Vec<BString>) -> Result<()> {
        self.header = Some(header);
        base.on_header(vec![b"row".into(), b"column".into(), b"key".into(), b"value".into()])
    }

    fn on_row(&mut self, base: &mut base::Base, mut row: Vec<BString>) -> Result<()> {
        if self.count == 0 {
            // first row
            if self.header.is_none() {
                base.on_header(vec![b"row".into(), b"column".into(), b"value".into()])?;
            }

        } else if self.opts.row_sep == base::AutoChoices::Always {
            base.on_separator()?;
        }

        self.count += 1;

        for (i, value) in row.drain(..).enumerate() {
            let mut row = vec![format!("{}", self.count).into(), format!("{}", i+1).into()];
            if let Some(header) = &self.header {
                if let Some(h) = header.get(i) {
                    row.push(format!("{h}").into());
                } else {
                    row.push(b"".into());
                }
            }
            row.push(value);
            base.on_row(row)?;
        }

        Break::when(self.opts.lines.is_some_and(|lines| self.count >= lines))
    }
}
